jira/resilientsession.py

```diff
@@ -100,20 +104,137 @@ def __init__(self, timeout=None):
         # Indicate our preference for JSON to avoid https://bitbucket.org/bspeakmon/jira-python/issue/46 and https://jira.atlassian.com/browse/JRA-38551
         self.headers.update({"Accept": "application/json,*.*;q=0.9"})
 
+    def _jira_prepare(self, **_kwargs) -> dict:
+        """Do any pre-processing of our own and return the updated kwargs."""
+        kwargs = _kwargs.copy()
+
+        request_headers = self.headers.copy()
+        request_headers.update(_kwargs.get("headers", {}))
+        kwargs["headers"] = request_headers
+
+        data = _kwargs.get("data", {})
+        if isinstance(data, dict):
+            # mypy ensures we don't do this,
+            # but for people subclassing we should preserve old behaviour
+            kwargs["data"] = json.dumps(data)
+
+        return kwargs
+
+    def request(  # type: ignore[override] # An intentionally different override
+        self,
+        method: str,
+        url: Union[str, bytes],
+        _prepare_retry_method: PrepareRequestForRetry = PassthroughRetryPrepare(),
+        **_kwargs,
+    ) -> Response:
+        """This is an intentional override of `Session.request()` to inject some error handling and retry logic.
+
+        Raises:
+            Exception: Various exceptions as defined in py:method:`raise_on_error`.
+
+        Returns:
+            Response: The response.
+        """
+
+        retry_number = 0
+        exception: Optional[Exception] = None
+        response: Optional[Response] = None
+        response_or_exception: Optional[Union[ConnectionError, Response]]
+
+        processed_kwargs = self._jira_prepare(**_kwargs)
+
+        def is_retrying() -> bool:
+            """Helper method to say if we should still be retrying."""
+            return retry_number <= self.max_retries
+
+        while is_retrying():
+            response = None
+            exception = None
+
+            try:
+                response = super().request(method, url, **processed_kwargs)
+                if response.ok:
+                    self.__handle_known_ok_response_errors(response)
+                    return response
+            # Can catch further exceptions as required below
+            except ConnectionError as e:
+                exception = e
+
+            # Decide if we should keep retrying
+            response_or_exception = response if response is not None else exception
+            retry_number += 1
+            if is_retrying() and self.__recoverable(
+                response_or_exception, url, method.upper(), retry_number
+            ):
+                _prepare_retry_method.prepare(processed_kwargs)
```

Comment:
I'm not too sure what this function is supposed to do: it returns the input unchanged but the returned value is ignored, am I missing something?

Reply:
yeah, the default does that, but in the case for adding an attachment we pass in a custom one which actually regenerates the data and headers for the request. 

Not an ideal situation, but happy to take alternatives, I went through a few iterations, but as this is where the retry behaviour and delays are this implementation has the least code!

Reply:
It's fine with me, I was asking out of curiosity because I couldn't figure it out! Consider this resolved.

Reply:
I'll leave this unresolved until I document the class!